sdk/webpubsub/azure-webpubsub-client/azure/webpubsub/client/_client.py

```diff
@@ -0,0 +1,858 @@
+                    ),
+                    **kwargs,
+                )
+            else:
+                self._send_message(
+                    message=SendToGroupMessage(group=group_name, data_type=data_type, data=content, no_echo=no_echo),
+                    **kwargs,
+                )
+
+        self._retry(send_to_group_attempt)
+
+    def _retry(self, func: Callable[[], None]):
+        retry_attempt = 0
+        while True:
+            try:
+                func()
+                return
+            except Exception as e:  # pylint: disable=broad-except
+                retry_attempt = retry_attempt + 1
+                delay_seconds = self._message_retry_policy.next_retry_delay(retry_attempt)
+                if delay_seconds is None:
+                    raise e
+                _LOGGER.debug("will retry %sth times after %s seconds", retry_attempt, delay_seconds)
+                time.sleep(delay_seconds)
+
+    def _call_back(self, callback_type: Union[CallbackType, str], *args):
+        for func in self._handler[callback_type]:
+            # _call_back works in listener thread which must not be blocked so we have to execute the func
+            # in new thread to avoid dead lock
+            threading.Thread(target=func, args=args, daemon=True).start()
+
+    def _start_from_restarting(self):
+        if self._state != WebPubSubClientState.DISCONNECTED:
+            _LOGGER.warning("Client can be only restarted when it's Disconnected")
+            return
+
+        try:
+            self._start_core()
+        except Exception as e:
+            self._state = WebPubSubClientState.DISCONNECTED
+            raise e
+
+    def _handle_auto_reconnect(self):
+        _LOGGER.debug("start auto reconnect")
+        success = False
+        attempt = 0
+        while not self._is_stopping:
+            try:
+                self._start_from_restarting()
+                success = True
+                break
+            except Exception as e:  # pylint: disable=broad-except
+                _LOGGER.warning("An attempt to reconnect connection failed %s", e)
+                attempt = attempt + 1
+                delay_seconds = self._reconnect_retry_policy.next_retry_delay(attempt)
+                if not delay_seconds:
+                    break
+                _LOGGER.debug("Delay time for reconnect attempt %d: %ds", attempt, delay_seconds)
+                time.sleep(delay_seconds)
+        if not success:
+            self._handle_connection_stopped()
+        else:
+            _LOGGER.debug("reconnect successfully")
+
+    def _handle_connection_stopped(self):
+        _LOGGER.debug("Connection stopped")
+        self._is_stopping = False
+        self._state = WebPubSubClientState.STOPPED
+        self._call_back(CallbackType.STOPPED)
+
+    def _handle_connection_close_and_no_recovery(self):
+        _LOGGER.debug("Connection closed and no recovery")
```

Comment:
What's the difference between "Connection stopped" (used above) and "Connection closed" (used here)?

Reply:
When "Connection stopped", client will be stopped.
When "Connection closed", client will try to reconnect or recover (configured when init). If success, it can be connected again; if fails, client will be stopped.